File: Source/Tests/CSharp/Transformation/FormulaTransformationTests.cs
﻿// The MIT License (MIT)
// 
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace Tests.CSharp.Transformation
{
	using System;
	using System.Collections.Immutable;
	using System.Reflection;
	using FluentAssertions;
	using NUnit.Framework;
	using SafetySharp.CSharp.Transformation;
	using SafetySharp.Formulas;
	using SafetySharp.Metamodel;
	using SafetySharp.Metamodel.Configurations;
	using SafetySharp.Metamodel.Declarations;
	using SafetySharp.Metamodel.Expressions;
	using SafetySharp.Metamodel.Types;
	using SafetySharp.Modeling;

	[TestFixture]
	internal class FormulaTransformationTests
	{
		[SetUp]
		public void Setup()
		{
			const string csharpCode = @"
				class X : Component
				{
					public int IntField;
					public bool BooleanField;
				}";

			_compilation = new TestCompilation(csharpCode);
			_assembly = _compilation.Compile();

			_symbolMap = new SymbolMap(_compilation.CSharpCompilation);
			_intFieldReference = _symbolMap.GetFieldReference(_compilation.FindFieldSymbol("X", "IntField"));
			_booleanFieldReference = _symbolMap.GetFieldReference(_compilation.FindFieldSymbol("X", "BooleanField"));

			_intFieldDeclaration = new FieldDeclaration(new Identifier("IntField"), TypeSymbol.Integer);
			_booleanFieldDeclaration = new FieldDeclaration(new Identifier("BooleanField"), TypeSymbol.Boolean);

			_componentResolver = ComponentResolver.Empty;
			_metamodelResolver = MetamodelResolver
				.Empty
				.With(_intFieldReference, _intFieldDeclaration)
				.With(_booleanFieldReference, _booleanFieldDeclaration);
		}

		private SymbolMap _symbolMap;
		private Assembly _assembly;
		private TestCompilation _compilation;
		private IMetamodelReference<FieldDeclaration> _intFieldReference;
		private IMetamodelReference<FieldDeclaration> _booleanFieldReference;
		private FieldDeclaration _intFieldDeclaration;
		private FieldDeclaration _booleanFieldDeclaration;
		private readonly FieldConfiguration _intFieldConfiguration1 = new FieldConfiguration(ImmutableArray.Create<object>(1, 2, 3));
		private readonly FieldConfiguration _booleanFieldConfiguration1 = new FieldConfiguration(ImmutableArray.Create<object>(true, false));
		private readonly FieldConfiguration _intFieldConfiguration2 = new FieldConfiguration(ImmutableArray.Create<object>(1, 2, 3));
		private readonly FieldConfiguration _booleanFieldConfiguration2 = new FieldConfiguration(ImmutableArray.Create<object>(true, false));
		private ComponentResolver _componentResolver;
		private MetamodelResolver _metamodelResolver;
		private FormulaResolver _formulaResolver;
		private Formula _transformedFormula;

		private Formula Transform(Formula formula)
		{
			var transformation = new FormulaTransformation(_compilation.ModelingCompilation, _symbolMap, _componentResolver, _metamodelResolver);
			_transformedFormula = transformation.Visit(formula);

			_formulaResolver = transformation.FormulaResolver;
			return _transformedFormula;
		}

		private StateFormula TransformStateFormula(string csharpExpression, params object[] values)
		{
			var untransformed = new UntransformedStateFormula(csharpExpression, values.ToImmutableArray());
			return (StateFormula)Transform(untransformed);
		}

		private Component CreateComponentInstance(string componentName, FieldConfiguration intField = null, FieldConfiguration booleanField = null)
		{
			intField = intField ?? _intFieldConfiguration1;
			booleanField = booleanField ?? _booleanFieldConfiguration1;
			var component = (Component)Activator.CreateInstance(_assembly.GetType(componentName));

			var fields = ImmutableDictionary<FieldDeclaration, FieldConfiguration>
				.Empty
				.Add(_intFieldDeclaration, intField)
				.Add(_booleanFieldDeclaration, booleanField);

			var configuration = new ComponentConfiguration(new Identifier("None"), ComponentDeclaration.Empty, fields,
														   ImmutableArray<ComponentConfiguration>.Empty);

			_componentResolver = _componentResolver.With(component.GetSnapshot(), configuration);
			return component;
		}

		private void CheckResolvedField(Expression fieldAccessExpression, FieldConfiguration fieldConfiguration)
		{
			ReferenceEquals(_formulaResolver.Resolve((FieldAccessExpression)fieldAccessExpression), fieldConfiguration).Should().BeTrue();
		}

		[Test]
		public void TransformComponentAccess()
		{
			var component1 = CreateComponentInstance("X", _intFieldConfiguration1, _booleanFieldConfiguration1);
			var component2 = CreateComponentInstance("X", _intFieldConfiguration2, _booleanFieldConfiguration2);
			var fieldAccess = new FieldAccessExpression(_booleanFieldReference);

			TransformStateFormula("{0}.BooleanField", component1)
				.Should().Be(new StateFormula(fieldAccess, null));

			CheckResolvedField(((StateFormula)_transformedFormula).Expression, _booleanFieldConfiguration1);

			TransformStateFormula("{0}.BooleanField == {1}.BooleanField", component1, component1)
				.Should().Be(new StateFormula(new BinaryExpression(fieldAccess, BinaryOperator.Equals, fieldAccess), null));

			CheckResolvedField(((BinaryExpression)((StateFormula)_transformedFormula).Expression).Left, _booleanFieldConfiguration1);
			CheckResolvedField(((BinaryExpression)((StateFormula)_transformedFormula).Expression).Right, _booleanFieldConfiguration1);
			
			TransformStateFormula("{0}.BooleanField == {1}.BooleanField", component1, component2)
				.Should().Be(new StateFormula(new BinaryExpression(fieldAccess, BinaryOperator.Equals, fieldAccess), null));

			CheckResolvedField(((BinaryExpression)((StateFormula)_transformedFormula).Expression).Left, _booleanFieldConfiguration1);
			CheckResolvedField(((BinaryExpression)((StateFormula)_transformedFormula).Expression).Right, _booleanFieldConfiguration2);
		}

		[Test]
		public void TransformInternalAccess()
		{
			var fieldAccess = new FieldAccessExpression(_intFieldReference);
			TransformStateFormula("{0} == 2", CreateComponentInstance("X").AccessInternal<int>("IntField"))
				.Should().Be(new StateFormula(new BinaryExpression(fieldAccess, BinaryOperator.Equals, new IntegerLiteral(2)), null));

			CheckResolvedField(((BinaryExpression)((StateFormula)_transformedFormula).Expression).Left, _intFieldConfiguration1);
		}

		[Test]
		public void TransformLiteralExpressions()
		{
			TransformStateFormula("true").Should().Be(new StateFormula(BooleanLiteral.True, null));
			TransformStateFormula("1 == 2")
				.Should().Be(new StateFormula(new BinaryExpression(new IntegerLiteral(1), BinaryOperator.Equals, new IntegerLiteral(2)), null));
			TransformStateFormula("1m == 2.5m")
				.Should().Be(new StateFormula(new BinaryExpression(new DecimalLiteral(1), BinaryOperator.Equals, new DecimalLiteral(2.5m)), null));
			TransformStateFormula("true || false")
				.Should().Be(new StateFormula(new BinaryExpression(BooleanLiteral.True, BinaryOperator.LogicalOr, BooleanLiteral.False), null));
		}

		[Test]
		public void TransformNestedFormula()
		{
			var booleanFieldAccess = new FieldAccessExpression(_booleanFieldReference);
			var intFieldAccess = new FieldAccessExpression(_intFieldReference);

			var fieldIsTrue = new UntransformedStateFormula("{0}.BooleanField", ImmutableArray.Create<object>(CreateComponentInstance("X")));
			var fieldIsTwo = new UntransformedStateFormula("{0}.IntField == 2", ImmutableArray.Create<object>(CreateComponentInstance("X")));

			var transformedFieldIsTrue = new StateFormula(booleanFieldAccess, null);
			var transformedfieldIsTwo = new StateFormula(new BinaryExpression(intFieldAccess, BinaryOperator.Equals, new IntegerLiteral(2)), null);

			Transform(new BinaryFormula(fieldIsTrue, BinaryTemporalOperator.Until, PathQuantifier.All, fieldIsTwo)).Should().Be(
				new BinaryFormula(transformedFieldIsTrue, BinaryTemporalOperator.Until, PathQuantifier.All, transformedfieldIsTwo));

			var leftExpression = ((StateFormula)((BinaryFormula)_transformedFormula).Left).Expression;
			var rightExpression = (BinaryExpression)((StateFormula)((BinaryFormula)_transformedFormula).Right).Expression;

			CheckResolvedField(leftExpression, _booleanFieldConfiguration1);
			CheckResolvedField(rightExpression.Left, _intFieldConfiguration1);
		}

		[Test]
		public void TransformValueAccess()
		{
			TransformStateFormula("{0}", true).Should().Be(new StateFormula(BooleanLiteral.True, null));
			TransformStateFormula("{1} == {0}", 2, 1)
				.Should().Be(new StateFormula(new BinaryExpression(new IntegerLiteral(1), BinaryOperator.Equals, new IntegerLiteral(2)), null));
			TransformStateFormula("{1} == {0}", 2m, 1.5m)
				.Should().Be(new StateFormula(new BinaryExpression(new DecimalLiteral(1.5m), BinaryOperator.Equals, new DecimalLiteral(2)), null));
			TransformStateFormula("{0} || {1}", true, false)
				.Should().Be(new StateFormula(new BinaryExpression(BooleanLiteral.True, BinaryOperator.LogicalOr, BooleanLiteral.False), null));
		}
	}
}

File: Source/Tests/CSharp/Transformation/FormulaTransformationTests.fs
// The MIT License (MIT)
// 
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.Tests.CSharp.Transformation.FormulaTransformationTests

open System
open System.Linq
open NUnit.Framework
open Swensen.Unquote
open SafetySharp.Internal.CSharp
open SafetySharp.Internal.Metamodel
open SafetySharp.Modeling
open SafetySharp.Tests.CSharp
open SafetySharp.Tests
open SafetySharp.Internal.CSharp.Transformation

[<AutoOpen>]
module private FormulaTransformationTestsHelper =
    let mutable symbolResolver = Unchecked.defaultof<SymbolResolver>
    let mutable objectResolver = Unchecked.defaultof<ObjectResolver>
    let mutable unknownComponent : IComponent = null

    let compile formulaType csharpCode =
        let compilation = TestCompilation (@"
            enum E { A, B, C }
            class A : Component { public bool boolField; private int intField; }
            class B : Component { public B(A a) { this.a = a; } private A a; public bool boolField; }
            class M : Model {
                public M() {
                    var a = new A();
                    var sub = new A();
                    var b = new B(sub);
                    Unknown = new A();

                    var enumValue = E.B;
                    int intValue = 17;
                    bool boolValue = true;

                    var accessIntField = a.Access<int>(""intField"");
                    var unknownAccess = Unknown.Access<bool>(""boolField"");

                    Formula = " + csharpCode + ";

                    SetPartitions(a, b);
                }

                public A Unknown;
                public " + formulaType + " Formula;
            }")
        
        let model = compilation.CreateObject<Model> "M"
        model.FinalizeMetadata ()
        symbolResolver <- SymbolTransformation.Transform compilation.CSharpCompilation model
        objectResolver <- ObjectTransformation.Transform model symbolResolver
        unknownComponent <- model.GetType().GetField("Unknown").GetValue model :?> Component
        model.GetType().GetField("Formula").GetValue model

    let transform formula =
        FormulaTransformation.Transform symbolResolver objectResolver formula

    let compileLtl csharpCode = 
        let formula = compile "LtlFormula" csharpCode
        transform (formula :?> LtlFormula).Formula

    let compileCtl csharpCode = 
        let formula = compile "CtlFormula" csharpCode
        transform (formula :?> CtlFormula).Formula

    let raisesUnknownComponent expression = 
        raisesWith<UnknownComponentException> expression (fun e -> <@ e.UnknownComponent = unknownComponent @>)

[<TestFixture>]
module ``Linear Temporal Logic`` =
    [<Test>]
    let ``throws when unknown component instance is accessed directly`` () =
        raisesUnknownComponent <@ compileLtl "Ltl.StateExpression(() => !Unknown.boolField)" @>

    [<Test>]
    let ``throws when unknown component instance is accessed indirectly`` () =
        raisesUnknownComponent <@ compileLtl "Ltl.StateExpression(() => !unknownAccess)" @>

    [<Test>]
    let ``transform state expression`` () =
        let oneEqualsSeventeen = BinaryExpression (IntegerLiteral 1, BinaryOperator.Equals, IntegerLiteral 17)
        let expected = StateFormula (BinaryExpression (BooleanLiteral true, BinaryOperator.LogicalOr, oneEqualsSeventeen))
        compileLtl "Ltl.StateExpression(() => true || 1 == intValue)" =? expected

    [<Test>]
    let ``transform enum literals in state expressions`` () =
        let expected = StateFormula (BinaryExpression (IntegerLiteral 1, BinaryOperator.Equals, IntegerLiteral 2))
        compileLtl "Ltl.StateExpression(() => enumValue == E.C)" =? expected

    [<Test>]
    let ``transform component indirect member access`` () =
        let actual = compileLtl "Ltl.StateExpression(() => accessIntField == intValue)"
        let fieldAccess = ReadField (symbolResolver.ComponentSymbols.[0].Fields.[1], Some symbolResolver.ModelSymbol.ComponentObjects.[0])
        let expected = StateFormula (BinaryExpression (fieldAccess, BinaryOperator.Equals, IntegerLiteral 17))
        actual =? expected

    [<Test>]
    let ``transform component direct member access`` () =
        let actual = compileLtl "Ltl.StateExpression(() => !a.boolField)"
        let fieldAccess = ReadField (symbolResolver.ComponentSymbols.[0].Fields.[0], Some symbolResolver.ModelSymbol.ComponentObjects.[0])
        let expected = StateFormula (UnaryExpression (fieldAccess, UnaryOperator.LogicalNot))
        actual =? expected

    [<Test>]
    let ``transform subcomponent direct member access`` () =
        let actual = compileLtl "Ltl.StateExpression(() => !sub.boolField)"
        let fieldAccess = ReadField (symbolResolver.ComponentSymbols.[0].Fields.[0], Some symbolResolver.ModelSymbol.ComponentObjects.[2])
        let expected = StateFormula (UnaryExpression (fieldAccess, UnaryOperator.LogicalNot))
        actual =? expected

    [<Test>]
    let ``transform mixed component member access`` () =
        let actual = compileLtl "Ltl.StateExpression(() => a.boolField != b.boolField)"
        let fieldAccessA = ReadField (symbolResolver.ComponentSymbols.[0].Fields.[0], Some symbolResolver.ModelSymbol.ComponentObjects.[0])
        let fieldAccessB = ReadField (symbolResolver.ComponentSymbols.[1].Fields.[0], Some symbolResolver.ModelSymbol.ComponentObjects.[1])
        let expected = StateFormula (BinaryExpression (fieldAccessA, BinaryOperator.NotEquals, fieldAccessB))
        actual =? expected

    [<Test>]
    let ``transform next`` () =
        let expression = BinaryExpression (IntegerLiteral 17, BinaryOperator.GreaterThan, IntegerLiteral 89)
        compileLtl "Ltl.Next(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.Next)

        compileLtl "Ltl.Next(Ltl.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.Next), UnaryFormulaOperator.Next)

    [<Test>]
    let ``transform finally`` () =
        let expression = BinaryExpression (IntegerLiteral 17, BinaryOperator.GreaterThan, IntegerLiteral 89)
        compileLtl "Ltl.Finally(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.Finally)

        compileLtl "Ltl.Finally(Ltl.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.Next), UnaryFormulaOperator.Finally)

    [<Test>]
    let ``transform globally`` () =
        let expression = BinaryExpression (IntegerLiteral 17, BinaryOperator.GreaterThan, IntegerLiteral 89)
        compileLtl "Ltl.Globally(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.Globally)

        compileLtl "Ltl.Globally(Ltl.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.Next), UnaryFormulaOperator.Globally)

    [<Test>]
    let ``transform until`` () =
        let stateFormula = StateFormula (BooleanLiteral true)
        let nextFormula = UnaryFormula (stateFormula, UnaryFormulaOperator.Next)

        compileLtl "Ltl.Until(() => boolValue, () => boolValue)" =? 
            BinaryFormula(stateFormula, BinaryFormulaOperator.Until, stateFormula)

        compileLtl "Ltl.Until(Ltl.Next(() => boolValue), () => boolValue)" =? 
            BinaryFormula(nextFormula, BinaryFormulaOperator.Until, stateFormula)

        compileLtl "Ltl.Until(() => boolValue, Ltl.Next(() => boolValue))" =? 
            BinaryFormula(stateFormula, BinaryFormulaOperator.Until, nextFormula)

        compileLtl "Ltl.Until(Ltl.Next(() => boolValue), Ltl.Next(() => boolValue))" =? 
            BinaryFormula(nextFormula, BinaryFormulaOperator.Until, nextFormula)

[<TestFixture>]
module ``Computation Tree Logic`` =
    [<Test>]
    let ``throws when unknown component instance is accessed directly`` () =
        raisesUnknownComponent <@ compileCtl "Ctl.StateExpression(() => !Unknown.boolField)" @>

    [<Test>]
    let ``throws when unknown component instance is accessed indirectly`` () =
        raisesUnknownComponent <@ compileCtl "Ctl.StateExpression(() => !unknownAccess)" @>

    [<Test>]
    let ``transform state expression`` () =
        let oneEqualsSeventeen = BinaryExpression(IntegerLiteral 1, BinaryOperator.Equals, IntegerLiteral 17)
        let expected =  StateFormula(BinaryExpression(BooleanLiteral true, BinaryOperator.LogicalOr, oneEqualsSeventeen))
        compileCtl "Ctl.StateExpression(() => boolValue || 1 == intValue)" =? expected

    [<Test>]
    let ``transform enum literals in state expressions`` () =
        let expected = StateFormula (BinaryExpression (IntegerLiteral 1, BinaryOperator.Equals, IntegerLiteral 2))
        compileCtl "Ctl.StateExpression(() => enumValue == E.C)" =? expected

    [<Test>]
    let ``transform component indirect member access`` () =
        let actual = compileCtl "Ctl.StateExpression(() => accessIntField == intValue)"
        let fieldAccess = ReadField (symbolResolver.ComponentSymbols.[0].Fields.[1], Some symbolResolver.ModelSymbol.ComponentObjects.[0])
        let expected = StateFormula (BinaryExpression (fieldAccess, BinaryOperator.Equals, IntegerLiteral 17))
        actual =? expected

    [<Test>]
    let ``transform component direct member access`` () =
        let actual = compileCtl "Ctl.StateExpression(() => !a.boolField)"
        let fieldAccess = ReadField (symbolResolver.ComponentSymbols.[0].Fields.[0], Some symbolResolver.ModelSymbol.ComponentObjects.[0])
        let expected = StateFormula (UnaryExpression (fieldAccess, UnaryOperator.LogicalNot))
        actual =? expected

    [<Test>]
    let ``transform subcomponent direct member access`` () =
        let actual = compileCtl "Ctl.StateExpression(() => !sub.boolField)"
        let fieldAccess = ReadField (symbolResolver.ComponentSymbols.[0].Fields.[0], Some symbolResolver.ModelSymbol.ComponentObjects.[2])
        let expected = StateFormula (UnaryExpression (fieldAccess, UnaryOperator.LogicalNot))
        actual =? expected

    [<Test>]
    let ``transform mixed component member access`` () =
        let actual = compileCtl "Ctl.StateExpression(() => a.boolField != b.boolField)"
        let fieldAccessA = ReadField (symbolResolver.ComponentSymbols.[0].Fields.[0], Some symbolResolver.ModelSymbol.ComponentObjects.[0])
        let fieldAccessB = ReadField (symbolResolver.ComponentSymbols.[1].Fields.[0], Some symbolResolver.ModelSymbol.ComponentObjects.[1])
        let expected = StateFormula (BinaryExpression (fieldAccessA, BinaryOperator.NotEquals, fieldAccessB))
        actual =? expected

    [<Test>]
    let ``transform next`` () =
        let expression = BinaryExpression (IntegerLiteral 17, BinaryOperator.GreaterThan, IntegerLiteral 89)
        compileCtl "Ctl.AllPaths.Next(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.AllPathsNext)

        compileCtl "Ctl.AllPaths.Next(Ctl.AllPaths.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.AllPathsNext), UnaryFormulaOperator.AllPathsNext)

        compileCtl "Ctl.ExistsPath.Next(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.ExistsPathNext)

        compileCtl "Ctl.ExistsPath.Next(Ctl.AllPaths.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.AllPathsNext), UnaryFormulaOperator.ExistsPathNext)

    [<Test>]
    let ``transform finally`` () =
        let expression = BinaryExpression (IntegerLiteral 17, BinaryOperator.GreaterThan, IntegerLiteral 89)
        compileCtl "Ctl.AllPaths.Finally(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.AllPathsFinally)

        compileCtl "Ctl.AllPaths.Finally(Ctl.AllPaths.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.AllPathsNext), UnaryFormulaOperator.AllPathsFinally)

        compileCtl "Ctl.ExistsPath.Finally(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.ExistsPathFinally)

        compileCtl "Ctl.ExistsPath.Finally(Ctl.AllPaths.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.AllPathsNext), UnaryFormulaOperator.ExistsPathFinally)

    [<Test>]
    let ``transform globally`` () =
        let expression = BinaryExpression (IntegerLiteral 17, BinaryOperator.GreaterThan, IntegerLiteral 89)
        compileCtl "Ctl.AllPaths.Globally(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.AllPathsGlobally)

        compileCtl "Ctl.AllPaths.Globally(Ctl.AllPaths.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.AllPathsNext), UnaryFormulaOperator.AllPathsGlobally)

        compileCtl "Ctl.ExistsPath.Globally(() => intValue > 89)" =? UnaryFormula(StateFormula expression, UnaryFormulaOperator.ExistsPathGlobally)

        compileCtl "Ctl.ExistsPath.Globally(Ctl.AllPaths.Next(() => boolValue))" =? 
            UnaryFormula(UnaryFormula(StateFormula (BooleanLiteral true), UnaryFormulaOperator.AllPathsNext), UnaryFormulaOperator.ExistsPathGlobally)

    [<Test>]
    let ``transform until`` () =
        let stateFormula = StateFormula (BooleanLiteral true)
        let nextFormula = UnaryFormula (stateFormula, UnaryFormulaOperator.AllPathsNext)

        compileCtl "Ctl.AllPaths.Until(() => boolValue, () => boolValue)" =? 
            BinaryFormula(stateFormula, BinaryFormulaOperator.AllPathsUntil, stateFormula)

        compileCtl "Ctl.AllPaths.Until(Ctl.AllPaths.Next(() => boolValue), () => boolValue)" =? 
            BinaryFormula(nextFormula, BinaryFormulaOperator.AllPathsUntil, stateFormula)

        compileCtl "Ctl.AllPaths.Until(() => boolValue, Ctl.AllPaths.Next(() => boolValue))" =? 
            BinaryFormula(stateFormula, BinaryFormulaOperator.AllPathsUntil, nextFormula)

        compileCtl "Ctl.AllPaths.Until(Ctl.AllPaths.Next(() => boolValue), Ctl.AllPaths.Next(() => boolValue))" =? 
            BinaryFormula(nextFormula, BinaryFormulaOperator.AllPathsUntil, nextFormula)

        compileCtl "Ctl.ExistsPath.Until(() => boolValue, () => boolValue)" =? 
            BinaryFormula(stateFormula, BinaryFormulaOperator.ExistsPathUntil, stateFormula)

        compileCtl "Ctl.ExistsPath.Until(Ctl.AllPaths.Next(() => boolValue), () => boolValue)" =? 
            BinaryFormula(nextFormula, BinaryFormulaOperator.ExistsPathUntil, stateFormula)

        compileCtl "Ctl.ExistsPath.Until(() => boolValue, Ctl.AllPaths.Next(() => boolValue))" =? 
            BinaryFormula(stateFormula, BinaryFormulaOperator.ExistsPathUntil, nextFormula)

        compileCtl "Ctl.ExistsPath.Until(Ctl.AllPaths.Next(() => boolValue), Ctl.AllPaths.Next(() => boolValue))" =? 
            BinaryFormula(nextFormula, BinaryFormulaOperator.ExistsPathUntil, nextFormula)